18F/doi-extractives-data
# Introduce a React version of the data explorer

Idea for refactoring mentioned in #871 

This is an exploratory commit to illustrate the first steps in
refactoring the data explorer into a React component. There are many
caveats to this PR. It only provides a minimal view, but one can imagine
that with a little bit of styling and extra code it could mirror the
existing table rows.

**Testing**
 - Navigate to the GDP explorer view
 - Scroll to the bottom of the existing table, there should be an unstyled table of rows with identical data from the existing table.
 - Open the filter pane and mess around with the sort-option (is this `$`/`%` toggle labelled wrong?) and the year. The data in the new table should update. Note that the select-a-region drop-down isn't supported in this PR.

**Caveats**
 - I have tried to minimize the site-wide impact of introducing React.
   This means, for example, that I have not introduced the JSX syntax.
   While JSX significantly improves the experience using React, it also
   requires some dependencies with webpack, such as Babel.

 - This is in no way supposed to be a feature-for-feature replacement.
   If the idea is valuable enough, it should be easy enough to cover all
   of the features, but it's not worth investing the time if this is a
   non-starter.

 - Some of the semantics of how I have put this together are the result
   of fitting React into the environment already provided. Using the
   location hash the way I do to get arguments is one example. In a full
   replacement, this would likely take a more elegant form.

 - Since I am less familiar with what general needs there are, I threw
   in some code that mixes concerns: hard-coding table column names and
   the like. This could easily be remedied with some generalization and
   is in fact a common pattern in React apps. For example, the
   `DataExplorer` renders `TableRow`s directly. To generalize, we would
   create some kind of `RegionalValuePercentRow` (with a much better
   name) and pass that as a prop into the `DataExplorer` as
   `rowRenderer` or similar.

 - I have also included several helper functions just because I don't know what already exists in the code and I don't know what is appropriate to reuse. See the data loader and parser, for example.

 - Against my preferences, I have limited this code strictly to ES5 syntax. Adding in fat-arrow functions and object/list destructuring would significantly improve the readability of portions of this PR.

 - For production-level introduction of React, the common libraries (such as React) can be bundled into a vendor JS bundle and the page-specific portions can continue to be loaded separately, so that for example, a page's front-matter could specify… `react: true` and the template would load the common bundle, while the `<script src="{{ site.base_url }}/js/pages/whatever.min.js"></script>` would remain just the way it is currently used.

cc: @shawnbot
